homeassistant/components/binary_sensor/zha.py

```diff
@@ -26,23 +30,48 @@
 
 async def async_setup_platform(hass, config, async_add_entities,
                                discovery_info=None):
-    """Set up the Zigbee Home Automation binary sensors."""
-    discovery_info = zha.get_discovery_info(hass, discovery_info)
+    """Old way of setting up Zigbee Home Automation binary sensors."""
+    pass
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Zigbee Home Automation binary sensor from config entry."""
+    async def async_discover(discovery_info):
+        await _async_setup_entity(hass, config_entry, async_add_entities,
+                                  discovery_info)
+
+    unsub = async_dispatcher_connect(
+        hass, ZHA_DISCOVERY_NEW.format(DOMAIN), async_discover)
+    hass.data[DATA_ZHA][DATA_ZHA_DISPATCHERS].append(unsub)
+
+    discovery_info = hass.data.get(DATA_ZHA, {})
+    binary_sensors = discovery_info.get('binary_sensor', None)
+    if binary_sensors is None:
+        return
+    for discovery_key in binary_sensors.keys():
+        await _async_setup_entity(hass, config_entry, async_add_entities,
```

Comment:
It's not very efficient to add one entity at a time. It's better to create all known entities for the platform first and then add all of them in one call to `async_add_entities`.

Reply:
> all of them in one call to `async_add_entities`.

I agree, but I think this would require quite extensive zha refactoring and would make this PR even larger. In current implementation, ZHA component doesn't keep the knowledge of all the entities in the network. Essentially it loads a ZHA node one at a time and then probes at platform level for entities. in my mind, it is not really a HUB, as platform doesn't "reach back" or talk to the "HUB" but operates directly at the platform level. So HUB is just responsible for kicking off the platform and then platform responsible for the entity operation. In other words: `[switch.entity] <--> [zha platform] <--> [zha network]`  and not `[switch.entity] <--> [zha platform] <--> [zha component] <--> [zha network]`

Reply:
The only change we need to make, as I see it, is to be able to pass a list of discovery info to `_async_setup_entity` and have it iterate that before calling `async_add_entities`.

Reply:
@MartinHjelmare what approach would you recommend to handle the following two scenarios:
- bulk entity add -- eg upon hass restart
- add a single entity -- eg a new ZHA device just joined the network?

we can make `_async_setup_entity` to iterate over discovery_info to create a list of entities and add them all via single `async_add_entities()` call.  But then at some point later a new ZHA device joins which should be represented by two additional sensor entities. Or should we  clean up discovery info once those entities were added?

Reply:
Yeah, since discovery_info is transient and not really needed to be stored permanently, it sounds good to clear the corresponding discovery_info after the entities have been added.

Does this solve the questions?

We should definitely make sure to handle both cases you describe.